sql: always update rows in merging indexes

fqazi opened the pull request:
Previously, when an index was merging we would optimize updates to avoid writing duplicate values. Once we added optimizations inside the MVCC merge to skip rows in the temporary index with newer timestamps we started noticing failures. This was because of the following scenario:

1) An index starts a backfill with data from t
2) An update modifies rows to set some value A in a column, this
   gets captured by the temporary index.
3) The backfill completes and the index is now merging.
4) An update modifies the same rows to set some value A, this update
   always gets captured by the temporary index, giving it a new newer
   timestamp. The new index that is being merged ignores the updates,
   because the primary index already has the same values.
5) During the merge we skip over value A because of the optimization
   introduced in #142768

To address this, this patch treats merging indexes like temporary indexes and always forces any updates into them. This acts as a transient state to ensure that any newer updates are guaranteed to hit both the temporary and  secondary index. Additionally, the index merger optimization is version gated, so that the queries changes are guaranteed to be there to make things safe.

Fixes: #143116

fqazi committed:
sql: always update rows in merging indexes

Previously, when an index was merging we would optimize updates to
avoid writing duplicate values. Once we added optimizations inside the
MVCC merge to skip rows in the temporary index with newer timestamps we
started noticing failures. This was because of the following scenario:

1) An index starts a backfill with data from t
2) An update modifies rows to set some value A in a column, this
   gets captured by the temporary index.
3) The backfill completes and the index is now merging.
4) An update modifies the same rows to set some value A, this update
   always gets captured by the temporary index, giving it a new newer
   timestamp. The new index that is being merged ignores the updates,
   because the primary index already has the same values.
5) During the merge we skip over value A because of the optimization
   introduced in #142768

To address this, this patch treats merging indexes like temporary
indexes and always forces any updates into them. This acts as a
transient state to ensure that any newer updates are guaranteed to hit
both the temporary and  secondary index. Additionally, the index merger
optimization is version gated, so that the queries changes are
guaranteed to be there to make things safe.

Fixes: #143116
Release note: None

spilchen reviewed (commented):
Good catch with this bug fix!

> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-:-OLtCLYX6eq-FVvkzEsz:b-463rr1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OLt8LIQ4J9-LXbudWVs:-OLt8LIQ4J9-LXbudWVt:b3kn2r9):*
Any way we can add a testcase to the DML injection tests for this scenario?
___
*[`pkg/sql/row/updater.go` line 426 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OLtARcb7FCO6jKNnYoE:-OLtARcc67xc-jq8j8tT:b6zzob8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc1d9804544644bc5a736d97716f984d26757adb/pkg/sql/row/updater.go#L426)):*
> ```Go
> 						//
> 						// For merging indexes we will compare timestamps during the merge
> 						// process so all updates should always be captured into the the
> ```

nit: spelling



_Suggestion:_
```Go
// process so all updates should always be captured into the
```


<!-- Sent from Reviewable.io -->

fqazi committed:
sql/schemachanger: update DML injection to insert during backfill

Previously, the DML injection test only injected at the start of every
phase. This meant that bugs involving DML in the middle of a bacfill /
merge could be missed. To address this, this patch adds a very basic set
of inserts / updates /deletes to ensure that index validation will not
fail.

Inform: #143116
Release note: None

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OLt8LIQ4J9-LXbudWVs:-OMD5Lhj1EQ378W25i_l:b-896fix):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Any way we can add a testcase to the DML injection tests for this scenario?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
Thanks for adding the test—merge doesn’t have great test coverage right now, so this is a helpful addition.

> Reviewed 7 of 7 files at r2, 84 of 84 files at r5, 4 of 4 files at r6, 80 of 80 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-:-OMDKBL4CDP2y8iPTwfW:b-d4lkjn)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`-- commits` line 57 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDINEJ2o_cpPDCkoxN:-OMDINEJ2o_cpPDCkoxO:bouxn0i):*
Did you mean to push this commit into this PR? It looks like a separate issue. Or was there a dependency here.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_index_and_column.go` line 69 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDK4SLE8-wWR-jn67L:-OMDK4SLE8-wWR-jn67M:b-5ar0fv) ([raw file](https://github.com/cockroachdb/cockroach/blob/480b33898ad7d7619aff8817d4a722b73cbb4df9/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_index_and_column.go#L69)):*
> ```Go
> 			relationID, columnID := rel.Var("table-id"), rel.Var("column-id")
> 			return rel.Clauses{
> 				from.Type((*scpb.PrimaryIndex)(nil)),
> ```

Just to clarify, this isn't a problem with secondary indexes? Only primary indexes?
___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 106 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDFdJ-AfCe2WPN1lLE:-OMDFdJ-AfCe2WPN1lLF:b8y8jn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c34c17ec25d125d8302dd1bcea8065969b53309/pkg/sql/schemachanger/dml_injection_test.go#L106)):*
> ```Go
> }
> 
> func (po phaseOrdinal) backfillDeleteRow(operationPO phaseOrdinal, isAfter bool) []string {
> ```

nit: these are values passed to the insert for the row to be deleted by the callback. Perhaps rename it to `backfillCallbackDeleteRow`?
___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDFmUf6GPZqaGckxuD:-OMDFmUf6GPZqaGckxuE:b3ylzo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c34c17ec25d125d8302dd1bcea8065969b53309/pkg/sql/schemachanger/dml_injection_test.go#L115)):*
> ```Go
> }
> 
> func (po phaseOrdinal) backfillUpdateRow(operationPO phaseOrdinal, isAfter bool) []string {
> ```

nit: similar comment here. Maybe rename `backfillCallbackUpdateRow`?
___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 569 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDE-VV921w_qsY7SNX:-OMDE-VV921w_qsY7SNY:b-w2xfyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c34c17ec25d125d8302dd1bcea8065969b53309/pkg/sql/schemachanger/dml_injection_test.go#L569)):*
> ```Go
> 			poCompleted := make(map[phaseOrdinal]struct{})
> 			var poSlice []phaseOrdinal
> 			// Seperate rows are updated and modified within the back fill.
> ```

nit: suggested rewording



_Suggestion:_
```Go
// Separate rows are updated and deleted within the backfill.
```
___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 572 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDEgzb7GD7zQWsU-oT:-OMDEgzb7GD7zQWsU-oU:bd6peb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c34c17ec25d125d8302dd1bcea8065969b53309/pkg/sql/schemachanger/dml_injection_test.go#L572)):*
> ```Go
> 			var currentPOForBackFill atomic.Value
> 			currentPOForBackFill.Store(phaseOrdinal{})
> 			var backfillBeforeDone, backfillAfterDone atomic.Bool
> ```

nit: I believe these control the INSERTs done in each stage of the backfill callback. Maybe suggest slightly different names such as: beforeBackfillCallbackDone and afterBackfillCallbackDone
___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 750 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDGesk2G3Xog6-p6ul:-OMDGesk2G3Xog6-p6um:b2lemv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c34c17ec25d125d8302dd1bcea8065969b53309/pkg/sql/schemachanger/dml_injection_test.go#L750)):*
> ```Go
> 				}
> 				// Execute queries to modify rows.
> 				operationUpdate := fmt.Sprintf("%s_%t", opBackfillUpdate, isAfter)
> ```

nit: I assume it's okay to fall in here if the callback has already been done? The DELETE or UPDATE should be no-ops.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 57 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDINEJ2o_cpPDCkoxN:-OME2HwbDTi9HLyEhy1M:bsc8t15):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Did you mean to push this commit into this PR? It looks like a separate issue. Or was there a dependency here.
</blockquote></details>

Done.

I'll split this off. Originally I thought this was dependency because the DML injection hit it. But I can't reproduce it now, I'll update the test later to hit it again. Also I think this is better dealt with in the optbuilder.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_index_and_column.go` line 69 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMDK4SLE8-wWR-jn67L:-OME2OLx9M8KNjAT6D2v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/480b33898ad7d7619aff8817d4a722b73cbb4df9/pkg/sql/schemachanger/scplan/internal/rules/current/dep_drop_index_and_column.go#L69)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Just to clarify, this isn't a problem with secondary indexes? Only primary indexes?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fqazi commented:
@spilchen RFAL

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 85 of 85 files at r8, 4 of 4 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-:-OMEsJqFEVtDW7vlqxwR:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/dml_injection_test.go` line 569 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/143267#-OMErr9E8xA2YF8p3U2A:-OMErr9E8xA2YF8p3U2B:b-3msotx) ([raw file](https://github.com/cockroachdb/cockroach/blob/81d0c61db55eb328e08f011c92f92ad9c6deb0fb/pkg/sql/schemachanger/dml_injection_test.go#L569)):*
> ```Go
> 			poCompleted := make(map[phaseOrdinal]struct{})
> 			var poSlice []phaseOrdinal
> 			// // Separate rows are updated and deleted within the backfill.
> ```

nit: extra `//` in the comment



_Suggestion:_
```Go
// Separate rows are updated and deleted within the backfill.
```


<!-- Sent from Reviewable.io -->

fqazi commented:
@spilchen @mgartner TFTR

bors r+
